mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,662 @@
+        q_net: Recurrent q network.
+        opt: Optimiser for the recurrent Q network.
+        rb: The replay buffer.
+
+    Returns:
+        The update function.
+    """
+
+    # INTERACT LEVEL 2
+    def select_eps_greedy_action(
+        action_selection_state: ActionSelectionState, obs: Observation, done: Array
+    ) -> Tuple[ActionSelectionState, Array]:
+        """Select action to take in epsilon-greedy way. Batch and agent dims are included.
+
+            Args:
+            action_selection_state: Tuple of online parameters, previous hidden state,
+                environment timestep (used to calculate epsilon) and a random key.
+            obs: The observation from the previous timestep.
+            done: The flag timestep.last() from the previous timestep.
+
+        Returns:
+            A tuple of the updated action selection state and the chosen action.
+        """
+        params, hidden_state, t, key = action_selection_state
+
+        eps = jax.numpy.maximum(
+            cfg.system.eps_min, 1 - (t / cfg.system.eps_decay) * (1 - cfg.system.eps_min)
+        )
+
+        obs = jax.tree_util.tree_map(lambda x: x[jnp.newaxis, ...], obs)
+        done = jax.tree_util.tree_map(lambda x: x[jnp.newaxis, ...], done)
+
+        next_hidden_state, eps_greedy_dist = q_net.apply(params, hidden_state, (obs, done), eps)
+
+        new_key, explore_key = jax.random.split(key, 2)
+
+        action = eps_greedy_dist.sample(seed=explore_key)
+        action = jnp.squeeze(action)
+
+        next_action_selection_state = ActionSelectionState(
+            params, next_hidden_state, t + cfg.arch.num_envs, new_key
+        )
+
+        return next_action_selection_state, action
+
+    # INTERACT LEVEL 1
+    def interaction_step(
+        interaction_state: InteractionState, _: Any
+    ) -> Tuple[InteractionState, Dict]:
+        """Selects action, steps global env, stores timesteps in rb and repacks the parameters."""
+        # light unpacking
+        action_selection_state, env_state, buffer_state, obs, done = interaction_state
+
+        # select the actions to take
+        next_action_selection_state, action = select_eps_greedy_action(
+            action_selection_state, obs, done
+        )
+
+        # step env with selected actions
+        next_env_state, next_timestep = jax.vmap(env.step)(env_state, action)
+
+        # Get reward
+        reward = next_timestep.reward
+        meaned_reward = jax.tree_util.tree_map(
+            lambda x: jnp.repeat(x[:, jnp.newaxis, ...], cfg.system.num_agents, axis=1),
+            jnp.mean(next_timestep.reward, axis=-1),
+        )
+
+        reward = jax.lax.select(
+            (jnp.zeros_like(reward) + cfg.system.mean_reward).astype(int), meaned_reward, reward
+        )
+
+        transition = Transition(obs, action, reward, done)
+        transition = jax.tree_util.tree_map(
+            lambda x: x[:, jnp.newaxis, ...], transition
+        )  # Add dummy time dim
```

Comment:
nit to avoid weird formatting. Also here if the new axis is time should it not be axis 0? As this is how you init your replay buffer?
```suggestion
        # Add dummy time dim
        transition = jax.tree_util.tree_map(lambda x: x[:, jnp.newaxis, ...], transition)  
```

Reply:
@sash-a This will be very important for using trajectorybuffers. This was where the biggest bug in my code was, and I'm gonna overexplain because this took me a while to uncover. I initialise the trajectory buffer with a single transition, so no need to worry about batch and time dimensions yet. The trajectory buffer takes in experience as B, T, ... 
(I quote from flashbax: 
" batch: A batch of experience. The leading axis of the pytree is the batch dimension.
            This must match `add_batch_size` and the structure of the experience used
            during initialisation of the buffer state. This batch is added along the time axis of
            the buffer state.")
And if you add a dummy time dimension as the first dimension, you will be effectively swapping the time and batch axes, and adding a batch of data along the time dimension, which totally messes up Q-learning.

I think you were looking at how I initialise my network, and there I use T, B, ... . I remember you helped me with this, it's necessary because the scannedRNN automatically scans over the first dimension, so it has to be time.

In conclusion, TrajectoryBuffer takes (B,T,...) while ScannedRNN (and therefore RecQNetwork) takes in (T,B, ...)

I hope I understood your comment correctly, and that it helps for recurrent SAC!

Reply:
Will resolve once you've seen, it will be NB!